crates/bevy_ecs/src/system/system_registry.rs

```diff
@@ -915,4 +967,25 @@ mod tests {
         world.run_system_cached(system.pipe(system)).unwrap();
         world.run_system_cached(system.map(|()| {})).unwrap();
     }
+
+    #[test]
+    #[should_panic]
+    fn removing_system_stack_with_one_shot() {
+        let mut world = World::new();
+        let _ = world.run_system_cached(|world: &mut World| {
+            world.remove_resource::<RunSystemStack>();
+        });
+    }
+
+    #[test]
+    #[should_panic]
+    fn removing_system_stack_with_observer() {
+        let mut world = World::new();
+        world.add_observer(
+            |_trigger: Trigger<OnRemove, RegisteredSystem<(), ()>>, world: &mut World| {
+                world.remove_resource::<RunSystemStack>();
```

Comment:
I think this is the cause of the miri failure.  It's not supposed to be valid to use an exclusive system as an observer!  I'm pretty sure it's a bug in the observer code that it allows this to be added.  Maybe we need to open an issue for that?

It should work to use `Commands` instead of `&mut World`, though.  

```suggestion
            |_trigger: Trigger<OnRemove, RegisteredSystem<(), ()>>, mut commands: Commands| {
                commands.remove_resource::<RunSystemStack>();
```